### PR TITLE
Individual exercise

### PR DESCRIPTION
**Pull request by lamont1008**

### CODE REVIEW

File: app/routes/exercise.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import express from 'express';
+import { ExerciseController } from '../http/controllers/ExerciseController';
+import { ExericseService } from '../services/ExerciseService';
```
Comment (JungAn2) [inline app/routes/exercise.ts:3]: You don't use exercise service here. Delete this import

File: app/http/controllers/ExerciseController.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { NextFunction, Request, Response } from "express";
+import { ExericseService } from "../../services/ExerciseService";
+
+export const ExerciseController = {
+
```
Comment (JungAn2) [inline app/http/controllers/ExerciseController.ts:5]: Add documentation on what this does

File: package.json [hunk 3]
```diff
@@ -42,7 +42,9 @@
     "axios": "^0.26.1",
     "dotenv": "^16.0.0",
     "express": "^4.17.2",
+    "front-matter": "^4.0.2",
```
Comment (JungAn2) [inline package.json:45]: We manually got front-matter so delete this library

File: tests/test-test.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+---
+title: this
+
+---
+
+# hello
+
+- bye
```
Comment (JungAn2) [inline tests/test-test.md:8]: Delete this file since we are not using this

File: app/services/ExerciseService.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
+import { marked } from 'marked'
+import NotFoundException from '../exceptions/NotFoundException';
+import fs from "fs";
+
+export const ExericseService = {
+
+  getContent(Exercise_ID: string) {
+
```
Comment (JungAn2) [inline app/services/ExerciseService.ts:8]: Add documentation for the function also delete the extra line below the function

File: app/services/ExerciseService.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      const form = marked.parse(fileContent)
+      var words = form.split("<hr>")
+      return words[0] + words[2]
+    } catch (e) {
+      throw new NotFoundException();
+    }
+
+  },
+
+  getMetaData(Exercise_ID: string): Object {
```
Comment (JungAn2) [inline app/services/ExerciseService.ts:21]: Add documentation

File: app/services/ExerciseService.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        metadata[eachCol[0]] = this.getDataType(eachCol[1]);
+      }
+      return metadata;
+    } catch (e) {
+      throw new NotFoundException()
+    }
+
+  },
+
+  getDataType(s: string): any {
```
Comment (JungAn2) [inline app/services/ExerciseService.ts:50]: Add documentation

File: app/services/ExerciseService.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    } else if (s.toLowerCase() == "true" || s.toLowerCase() == "false") {
+      return Boolean(s);
+    } else {
+      // Gets rid of double quotation
+      s = s.substring(1, s.length - 1);
+      return s;
+    }
+  },
+
+  getDisplay(exercise_ID: string) {
```
Comment (JungAn2) [inline app/services/ExerciseService.ts:74]: Add documentation

**Review by JungAn2 — CHANGES_REQUESTED** [review]
Couple changes with not using library and imports.
Also need to document functions

example
``` ts
/**
* returns given message
*
* @param {string} message
* @return {string} the message passed in
*/
 getMessage(message: string){
  return message;
}
```

**Review by JungAn2 — APPROVED** [review]
Looks good

File: app/http/controllers/ExerciseController.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+export const ExerciseController = {
+    /**
+      * Handles the login request
+      * 
+      * @param req {Request} Express request object
+      * @param res {Response} Express response object
+      * @param next {NextFunction} Express NextFunction (used for middleware)
+      * @throws {NotFoundException} File is Not found
+      */
+    getInfo(req: Request, res: Response, next: NextFunction) {
```
Comment (wrathofrathma) [inline app/http/controllers/ExerciseController.ts:13]: The exercise controller's "getInfo" controller handles the login request?

File: app/services/ExerciseService.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+* @return {string} the content being returned
+* @throws {NotFoundException} File is Not found exception handler
+*/
+  getContent(Exercise_ID: string): string {
+    const fileLocation = "exercises/" + Exercise_ID + ".md";
+    try {
+      const fileContent = fs.readFileSync(fileLocation, "utf8");
+      const form = marked.parse(fileContent)
+      var words = form.split("<hr>")
+      return words[0] + words[2]
```
Comment (wrathofrathma) [inline app/services/ExerciseService.ts:19]: This is a pretty perfect area to get in the habit of commenting. At the moment I'm not sure why we're splitting by "<hr>", and in a year from now you might not remember exactly either. Nor what the contents of words[0] and words[2] are. 

Can you add some descriptive comments to shed light on those two things?

File: app/services/ExerciseService.ts [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    try {
+      const fileContent = fs.readFileSync(fileLocation, "utf8");
+      const lexer = marked.lexer(fileContent);
+      let content = "";
+      for (let i = 0; i < lexer.length; i++) {
+        if (lexer[i].type == "hr") {
+          content = lexer[i + 1].text;
+          break;
+        }
+      }
```
Comment (wrathofrathma) [inline app/services/ExerciseService.ts:44]: Comments on the goal of this for loop / what lexer[i+1] is.

File: app/services/ExerciseService.ts [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        split[i] = split[i].substring(1, split[i].length - 1);
+      }
+      return split;
+    } else if (s.toLowerCase() == "true" || s.toLowerCase() == "false") {
+      return Boolean(s);
+    } else {
+      // Gets rid of double quotation
+      s = s.substring(1, s.length - 1);
+      return s;
+    }
```
Comment (wrathofrathma) [inline app/services/ExerciseService.ts:89]: This is okay for now if the tests are passing. 

In the future we should explore either a library to handle this heavy lifting. (All of these types are builtin, so I imagine there's something we can do)

File: app/services/ExerciseService.ts [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+      display["content"] = content
+      display["metadata"] = metadata
+
+      return JSON.parse(JSON.stringify(display))
+    } catch (e) {
+      throw new NotFoundException()
+    }
+
+  }
```
Comment (wrathofrathma) [inline app/services/ExerciseService.ts:113]: Kind of a weird name for the function. The content of this is fine, but we're not really displaying anything. We're fetching / getting it. 

Refactor the name of the function / description to be something more accurate and it's good.

**Review by wrathofrathma — CHANGES_REQUESTED** [review]
Just a few requested changes. No logical changes, just some comments / naming things I think.

Also I'm not sure when we got into the habit of doing upper camel case snake case for arguments and constant upper case for some variables, but that'll likely be changed over the summer and enforced against next semester. 

Examples of what I mean for both

In several functions in the ExerciseService, we are passing the argument "Exercise_ID", which should be "exerciseID" if we go by google's typescript guide(lower camel case), or exercise_id, if we internally decide on snake case. 

In the AuthService, we define a few variables "PAYLOAD" and TOKEN_HEADER. Which should be in lower camel case or lower snake case(much like above, depending on what we decide). I'd say lower camel case is a good standard to go with. The only time we use constant upper case would be global constants / enums.

**Review by wrathofrathma — APPROVED** [review]
Good improvements on the commenting. 

Happy to see we're already fixing the casing of our variables as well!

Looks good to me.

**Pull request merged**